service/internal/hamaster/ha_master.go

```diff
@@ -0,0 +1,200 @@
+package hamaster
+
+import (
+	"context"
+	"fmt"
+
+	infrastructurev1alpha2 "github.com/giantswarm/apiextensions/pkg/apis/infrastructure/v1alpha2"
+	"github.com/giantswarm/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/cachekeycontext"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/aws-operator/pkg/label"
+	"github.com/giantswarm/aws-operator/service/controller/key"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+}
+
+type HAMaster struct {
+	k8sClient k8sclient.Interface
+
+	awsCache map[string]infrastructurev1alpha2.AWSControlPlane
+	g8sCache map[string]infrastructurev1alpha2.G8sControlPlane
+}
+
+func New(config Config) (*HAMaster, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+
+	h := &HAMaster{
+		k8sClient: config.K8sClient,
+
+		awsCache: map[string]infrastructurev1alpha2.AWSControlPlane{},
+		g8sCache: map[string]infrastructurev1alpha2.G8sControlPlane{},
+	}
+
+	return h, nil
+}
+
+func (h *HAMaster) Mapping(ctx context.Context, obj interface{}) ([]Mapping, error) {
+	var err error
+	var ok bool
+
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var cacheKey string
+	{
+		ck, ok := cachekeycontext.FromContext(ctx)
+		if ok {
+			cacheKey = fmt.Sprintf("%s/%s", ck, key.ClusterID(cr))
+		}
+	}
```

Comment:
Note that I added local caching. See also https://github.com/giantswarm/operatorkit/pull/390.